src/Voron/Data/Compression/DecompressionBuffersPool.cs

```diff
@@ -303,23 +303,59 @@ public DecompressionBuffer(AbstractPager pager, long position, int size, Decompr
 
             public void EnsureValidPointer(LowLevelTransaction tx)
             {
-                _pager.EnsureMapped(tx, _position, _size / Constants.Storage.PageSize);
-                var p = _pager.AcquirePagePointer(tx, _position);
+                PagerInfo.Pager.EnsureMapped(tx, _position, _size / Constants.Storage.PageSize);
+                var p = PagerInfo.Pager.AcquirePagePointer(tx, _position);
 
                 TempPage.SetPointer(p);
             }
 
-            public bool CanReuse => _pager.Disposed == false;
-
             public void Dispose()
             {
-                if (_pager.Options.Encryption.IsEnabled)
+                if (PagerInfo.Pager.Options.Encryption.IsEnabled)
                     Sodium.sodium_memzero(TempPage.TempPagePointer, (UIntPtr)TempPage.PageSize);
 
                 // return it to the pool
                 _pool._pool[_index].Enqueue(this);
 
                 Interlocked.Add(ref _pool._currentlyUsedBytes, -_size);
+                PagerInfo.Release();
+            }
+        }
+
+        private class PagerInfo
+        {
+            internal readonly AbstractPager Pager;
+            private long _numberOfUsages;
+
+            public PagerInfo(AbstractPager pager)
+            {
+                Pager = pager;
+            }
+
+            public bool TryUse()
+            {
+                return Interlocked.Increment(ref _numberOfUsages) > 0;
+            }
+
+            public void Release()
+            {
+                Interlocked.Decrement(ref _numberOfUsages);
+            }
+
+            public bool TryDispose()
+            {
+                if (Interlocked.Read(ref _numberOfUsages) > 0)
+                    return false;
+
+                if (Interlocked.CompareExchange(ref _numberOfUsages, -(1000 * 1000), 0) != 0)
+                    return false;
+
+                return true;
+            }
+
+            public void Dispose()
```

Comment:
Given the `Dispose()` method then this class should implement `IDisposable`. BUT this is just a "helper" class / kind of container so maybe renaming this method to `DisposePager()` will be more suitable here.

Reply:
done